src/Angor/Avalonia/Angor.Wallet/Infrastructure/Impl/SensitiveWalletDataProvider.cs

```diff
@@ -1,22 +1,36 @@
 using Angor.Wallet.Domain;
 using Angor.Wallet.Infrastructure.Interfaces;
 using CSharpFunctionalExtensions;
+using CSharpFunctionalExtensions.ValueTasks;
 
 namespace Angor.Wallet.Infrastructure.Impl;
 
 public class SensitiveWalletDataProvider(IWalletStore walletStore, IWalletSecurityContext walletSecurityContext) : ISensitiveWalletDataProvider
 {
+    private readonly Dictionary<WalletId, (string, string)> sensitiveData = new();
+
+    public async Task<Result<(string seed, string passphrase)>> RequestSensitiveData(WalletId id)
+    {
+        var findResult = sensitiveData.TryFind(id);
+        if (findResult.HasValue)
+        {
+            return findResult.Value;
+        }
+
+        return await RequestSensitiveDataCore(id).Tap(data => sensitiveData[id] = data);
+    }
+
     private async Task<Result<EncryptedWallet>> GetEncryptedWallet(WalletId id)
     {
         return await walletStore.GetAll()
             .Map(list => list.FirstOrDefault(x => x.Id == id.Id))
             .EnsureNotNull("Wallet not found");
     }
-    
-    public async Task<Result<(string seed, string passphrase)>> RequestSensitiveData(WalletId id)
+
+    private async Task<Result<(string seed, string passphrase)>> RequestSensitiveDataCore(WalletId id)
     {
         return await GetEncryptedWallet(id)
-            .Bind(encryptedWallet => 
+            .Bind(encryptedWallet =>
                 walletSecurityContext.EncryptionKeyProvider.Get(id).ToResult("Encryption key not provided")
```

Comment:
IMHO there are too many methods here called without breaking line so it becomes really hard to see the whole flow, especially when some are called internally, I would prefer break line for each method call or add parameters and not chain everything..

Reply:
Done!